### ISSUE TITLE
Links Not Transitioning Properly: sample_image

### ISSUE THREAD

**Issue by purvigoel** [wphicks/Writing3D#25]
In sample_image, clicking a link causes a movement of the picture but the respective link does not disappear in favor of the next link, despite the following code in sample_image.xml:

 Actions
                  ObjectChange name="frontlink"
                           Transition duration="1.0"
                           Visible false /Visible 
                           /Transition
                  /ObjectChange
/Actions

which ideally should be making "front link" disappear. I believe this is one of the issues causing the odd, stunted movement of the picture, as the link "Fly Front" is still active when the user wants the image to "Fly Right" or "Fly Left"

**Comment by purvigoel** [comment]
A very similar problem occurs in sample_model. When rotations and movement are turned on in the xml file, the sample works properly except for the above commands ("Fly Front", "Fly Right," etc), which do not disappear after being clicked. I am not sure if this indicates a problem in the rotation mechanisms.

**Comment by wphicks** [comment]
Ah, excellent. My suspicion is that this is unrelated to the rotation, but I'm having trouble reproducing the bug for unrelated reasons. Specifically, I think we may have different versions of the legacy samples.

Would you mind zipping up just your samples directory and e-mailing it to me? I'll then take the important samples, modify them slightly if necessary and post them to the repo. That way, future contributors will be able to see what we're talking about, and we'll be guaranteed to be on the same page.

**Comment by wphicks** [comment]
Got the samples and was able to reproduce the bug. I realized that we need to get permission from the people who created those samples ( @ikreymer, I think) before we post them. We should see if we can get them covered by either the GPL license we're using for the entire project or a CC license of some sort.

I haven't figured out exactly what's causing the bug yet, but my most recent commit should help track it down. I've just added logging that can be switched on by adding the following sub-node to the Global node of the project xml:
`<Debug>true</Debug>`
Feel free to start liberally sprinkling `W3D_LOG.debug` calls into any bge code to help track down whatever you're currently working on.

**Comment by purvigoel** [comment]
Just thought I'd log down where I'm at with this issue right now! 

I used debug statements for visibility and got the following whenever clicking a link:

2016-07-05 09:02:21,393    DEBUG W3D object visibility boolean set to False
2016-07-05 09:02:21,393    DEBUG W3D object visibility boolean set to True

This seems to be correct--the visibility of one link is turned off, the visibility of the subsequent link is turned on. 
I also used print lines to see how the initial state of the links were set up, and the following is an excerpt from "right link" 

``` python
{'object_name': 'rightlink', 'duration': 1.0, 'visible': True}
            blender_object.color[3] = int(blender_object.visible)
            blender_object.setVisible(True)
            delta_alpha = 1 - blender_object.color[3]
            W3D_LOG.debug('object visibility boolean set to True')
            blender_object['visV'] = delta_alpha/(1.0*bge.logic.getLogicTicRate())
            new_color = blender_object.color
            new_color[3] = 1
            blender_object.color = new_color
            blender_object.setVisible(True)
            W3D_LOG.debug('object visibility set to True at end')
{'object_name': 'rightlink', 'duration': 1.0, 'visible': False}
            blender_object.color[3] = int(blender_object.visible)
            blender_object.setVisible(True)
            delta_alpha = 0 - blender_object.color[3]
            W3D_LOG.debug('object visibility boolean set to False')
            blender_object['visV'] = delta_alpha/(1.0*bge.logic.getLogicTicRate())
            new_color = blender_object.color
            new_color[3] = 0
            blender_object.color = new_color
            blender_object.setVisible(False)
            W3D_LOG.debug('object visibility set to False at end')
```

This seems to be correct as well--the link is first turned "on" , then turned "off". Eventually, every link but the first link is turned "on".

I did notice however in the debug statements checking status, I generally got the following when clicking the first "front link". 

2016-07-05 09:02:21,410    DEBUG W3D activate called on link_object_floorlink with status Continue
2016-07-05 09:02:21,432    DEBUG W3D activate called on link_object_floorlink with status Continue
2016-07-05 09:02:21,443    DEBUG W3D activate called on link_object_floorlink with status Continue
2016-07-05 09:02:21,460    DEBUG W3D activate called on link_object_floorlink with status Continue
2016-07-05 09:02:21,484    DEBUG W3D activate called on link_object_floorlink with status Continue
2016-07-05 09:02:21,493    DEBUG W3D activate called on link_object_floorlink with status Continue
2016-07-05 09:02:21,510    DEBUG W3D activate called on link_object_floorlink with status Continue
2016-07-05 09:02:21,531    DEBUG W3D activate called on link_object_floorlink with status Continue
2016-07-05 09:02:21,543    DEBUG W3D activate called on link_object_floorlink with status Continue
2016-07-05 09:02:21,561    DEBUG W3D activate called on link_object_floorlink with status Continue
2016-07-05 09:02:21,581    DEBUG W3D activate called on link_object_floorlink with status Continue
2016-07-05 09:02:21,593    DEBUG W3D activate called on link_object_floorlink with status Continue
2016-07-05 09:02:21,611    DEBUG W3D activate called on link_object_floorlink with status Continue
2016-07-05 09:02:21,626    DEBUG W3D activate called on link_object_floorlink with status Continue
2016-07-05 09:02:21,643    DEBUG W3D activate called on link_object_floorlink with status Continue

However, "floorlink" was not even an option in the visualization; the option after clicking "Fly Front" is "Fly Right," and "Fly Floor" does not appear until after that. This activation of floorlink was returned after just clicking "front link." I wasn't sure if this was a problem, so I'm just noting it down here. 

I also noticed that in sample_image.xml, frontlink and leftlink have the "Remain Enabled" option set to false, while rightlink and floorlink have the "Remain Enabled" option set to true. I tried fiddling with it and didn't see a huge difference, but again I wasn't sure if this was a problem.

**Comment by purvigoel** [comment]
I had a little more luck when looking at sample.xml. There was a similar problem here; the links wouldn't disappear/reappear at the correct time. However, I noticed another problem which may or may not be related: the moving text would often move to the incorrect wall. This might be what's throwing off the readings when the object has to move to the next wall. 

Upon initialization, the following is traced: 

2016-07-06 10:03:04,663    DEBUG W3D activate called on link_object_leftlink with status Stop
2016-07-06 10:03:04,664    DEBUG W3D activate status stopped on link_object_leftlink
2016-07-06 10:03:04,666    DEBUG W3D activate called on link_object_floorlink with status Stop
2016-07-06 10:03:04,667    DEBUG W3D activate status stopped on link_object_floorlink
2016-07-06 10:03:04,670    DEBUG W3D activate called on link_object_rightlink with status Stop
2016-07-06 10:03:04,670    DEBUG W3D activate status stopped on link_object_rightlink
2016-07-06 10:03:04,673    DEBUG W3D activate called on link_object_frontlink with status Stop
2016-07-06 10:03:04,673    DEBUG W3D activate status stopped on link_object_frontlink

Now the first option is "Fly Front", so we would expect upon choosing this that we would get the line:
DEBUG W3D activate called on link_object_frontlink with status Continue.

This does actually happen--but only about 1/10 of the time. More often, we get:

2016-07-06 10:03:07,976    DEBUG W3D activate called on link_object_leftlink with status Start
2016-07-06 10:03:07,993    DEBUG W3D activate called on link_object_leftlink with status Continue
2016-07-06 10:03:07,993    DEBUG W3D Position is LeftWall

Sometimes, this happens instead with link_object_floorlink or link_object_rightlink. Essentially, it seems like the wrong link is being activated, despite the fact that only one link can be clicked.

**Comment by wphicks** [comment]
Hmmm... Very interesting... Can you commit your logging version, so I can try to trace the exact chain of events here? I'm especially interested in the fact that we get the expected action about 1/10th of the time. That sounds to me like a timing issue of some sort.

**Comment by purvigoel** [comment]
Sure, just sent in the commit!

**Comment by wphicks** [comment]
Discovered root cause here was that in legacy software, invisible links cannot be clicked, whereas Writing3D allowed clicking of visible links (credit to @purvigoel for figuring this out). Will pull in @purvigoel 's bugfix shortly. Note that testVisibility.xml (also to be pulled in from @purvigoel) is a simple test for this issue. Will close this issue once changes are pulled in; please open new issues if other problems are discovered for this sample.

**Comment by wphicks** [comment]
Resolved by 5508853b9bdc0d248974b0c81013c13547c73a89
